### DISCUSSION TITLE
Prone Sniper Rifle Shooting

### DISCUSSION THREAD

**Discussion by blackfirefly000** [alberttheprince/rpemotes-reborn#40] [Ideas]
This discussion is in regards to a [feature request]( https://forum.cfx.re/t/free-rpemotes-reborn-a-standalone-emote-system-for-fivem/5219460/100) by [porterslug](https://forum.cfx.re/u/porterslug) on the fivem forums for prone sniper shooting.

I did a bit of coding on this a few weeks ago but was unsuccessful and deleted that script, but I've decided to take another crack at it.

Given that the main trouble comes from the vanilla sniper rifle aim and shooting animations will overwrite any animation the player is in, this solution attempts to freeze the player every frame to prevent that. It doesn't work. I've included the resource in a .zip file below so anyone can take a look at it as a jumping-off point. Note that this script was developed in part with the help of Chat GPT.
[sniper.zip](https://github.com/user-attachments/files/16593542/sniper.zip)

client.lua:
```
RegisterCommand("sniper", function(source, args, rawCommand)
    local animationDict = "smo@tactical_53"
    local animationName = "tactical_53_clip"

    -- Load the animation dictionary
    RequestAnimDict(animationDict)

    while not HasAnimDictLoaded(animationDict) do
        Wait(100) -- Wait for the animation dictionary to load
    end

    local playerPed = PlayerPedId()

    -- Play the animation
    TaskPlayAnim(playerPed, animationDict, animationName, 8.0, 8.0, -1, 32, 0, false, false, false)

    -- Enable freeze every frame
    isAnimating = true
    Citizen.CreateThread(function()
        while isAnimating do
            Citizen.Wait(0)

            -- Freeze player
            FreezeEntityPosition(playerPed, true)
            SetEntityCollision(playerPed, false, false)

            -- Disable movement controls
            DisableControlAction(0, 30, true)  -- Disable moving forward/backward
            DisableControlAction(0, 31, true)  -- Disable moving left/right
        end
    end)
end, false)

RegisterCommand("stopsniper", function(source, args, rawCommand)
    local playerPed = PlayerPedId()

    -- Stop the animation and unfreeze the player
    ClearPedTasksImmediately(playerPed)
    FreezeEntityPosition(playerPed, false)
    SetEntityCollision(playerPed, true, true)

    -- Re-enable movement
    EnableControlAction(0, 30, true)
    EnableControlAction(0, 31, true)

    -- Stop freeze
    isAnimating = false
end, false)
```

**Comment by alberttheprince** [comment]
@MadsLeander Would you have any insight into this considering you've worked on crawling?

**Reply by MadsLeander** [reply]
I'll take a look at this entire thread sometime tomorrow and give my thoughts on it.

**Comment by iSentrie** [comment]
Standing, overriding animations isn't the only issues, next one is desync (you will be laying down, but for others standing), that's why it was disabled and kept that way. If i remember correctly.

**Reply by alberttheprince** [reply]
The idea here is to freeze the player's model and force them into the animation while they shoot. I'm not sure if this is possible; it may need to be done every frame.

It would not be performance friendly I'm sure, but it would be good to even get to that point.

**Reply by alberttheprince** [reply]
Or would this require making something to basically "sync" this between players? I've seen some people do this with some things like:

https://github.com/Manason/mana_audio

Which syncs audio generally played client side only.

**Reply by blackfirefly000** [reply]
> The idea here is to freeze the player's model and force them into the animation while they shoot. I'm not sure if this is possible; it may need to be done every frame.
> 
> It would not be performance friendly I'm sure, but it would be good to even get to that point.

I've completely rewrote the script, roughly following [Andyyy7666](https://github.com/Andyyy7666)/[WeaponAnimationMenuV2](https://github.com/Andyyy7666/WeaponAnimationMenuV2) and that's not necessary. The problem is that the prone sniper animation dictionaries are designed to be static. In order to work for aiming or firing, they need the following clips/clipsets/animations: fire_high, fire_low, fire_med, grip_med, grip_low, fire_recoil, and grip_high. (Potentially they need more). The 2 prone sniper animations/dictionaries I've found ([Firearms Poses](https://www.gta5-mods.com/misc/firearms-poses), [Custom Tactical Poses](https://www.gta5-mods.com/misc/custom-rifle-poses-add-on)) don't include those 

Rewrite below:
Note that 1: the animation is still broken. See above
2: isPlayerUsingSniperRifle(ped) isn't currently working and is commented out
[sniper0.2.zip](https://github.com/user-attachments/files/16607540/sniper0.2.zip)

```
    local animationDict = "smo@tactical_53" -- Replace with your animation dictionary
    local animationName = "tactical_53_clip" -- Replace with your animation name

    -- List of sniper rifle weapon hashes
    local sniperRifles = {
    "WEAPON_SNIPERRIFLE",
    "WEAPON_HEAVYSNIPER",
    "WEAPON_MARKSMANRIFLE",
    "WEAPON_HEAVYSNIPER_MK2",
    "WEAPON_MARKSMANRIFLE_MK2",
    "weapon_precisionrifle"
}

CreateThread(function()
    while true do
        Citizen.Wait(0)
        ped = PlayerPedId()
        inVeh = IsPedInVehicle(PlayerPedId(-1), GetVehiclePedIsIn(PlayerPedId(-1), false), false)
        local _, hash = GetCurrentPedWeapon(Player, 1)
        if not inVeh then
            loadAnimDict(animationDict)
            if IsPlayerFreeAiming(ped) --[[and isPlayerUsingSniperRifle(ped)--]] or (IsControlPressed(0, 24) --[[and isPlayerUsingSniperRifle(ped)--]] and GetAmmoInClip(ped, hash) > 0) then
                if not IsEntityPlayingAnim(ped, animationDict, animationName, 3) then
                    TaskPlayAnim(ped, animationDict, animationName, 8.0, -8.0, -1, 32, 0, 0, 0, 0)
                    DisableControlAction(0, 30, true)  -- Disable moving forward/backward
                    DisableControlAction(0, 31, true)  -- Disable moving left/right
                end
            elseif IsEntityPlayingAnim(ped, animationDict, animationName, 3) then
                ClearPedTasks(ped)
                EnableControlAction(0, 30, true)
                EnableControlAction(0, 31, true)
            end
            Citizen.Wait(50)
        end
        Citizen.Wait(50)
    end
end)

function loadAnimDict(dict)
	while (not HasAnimDictLoaded(dict)) do
		RequestAnimDict(dict)
		Citizen.Wait(50)
	end
end

-- Function to check if the player is using a sniper rifle
function isPlayerUsingSniperRifle(ped)
    local currentWeapon = GetSelectedPedWeapon(ped) -- Get the weapon hash of the current weapon

    if IsEntityDead(ped) then
		blocked = false
			return false
		else
            for _, weaponHash in ipairs(sniperRifles) do
                if currentWeapon == GetHashKey(weaponHash) then
                    return true
                end
            end
        return false
    end
end
```

Animations used:
[Firearms Poses](https://www.gta5-mods.com/misc/firearms-poses)
Alternate: [Custom Tactical Poses](https://www.gta5-mods.com/misc/custom-rifle-poses-add-on)

**Comment by blackfirefly000** [comment]
Did some code rewrites (it's still not working quite right)

```
    local animationDict = "smo@tactical_53" -- Replace with your animation dictionary
    local animationName = "tactical_53_clip" -- Replace with your animation name

CreateThread(function()
    while true do
        Citizen.Wait(0)
        ped = PlayerPedId()
        inVeh = IsPedInVehicle(PlayerPedId(-1), GetVehiclePedIsIn(PlayerPedId(-1), false), false)
        if not inVeh then
            loadAnimDict(animationDict)
            if (GetWeapontypeGroup(GetSelectedPedWeapon(ID)) == GetWeapontypeGroup(`WEAPON_SNIPER`)) and (IsPlayerFreeAiming(ped) or IsPedShooting(ped)) then
                if not IsEntityPlayingAnim(ped, animationDict, animationName, 3) then
                    TaskPlayAnim(ped, animationDict, animationName, 8.0, -8.0, -1, 32, 0, 0, 0, 0)
                    DisableControlAction(0, 30, true)  -- Disable moving forward/backward
                    DisableControlAction(0, 31, true)  -- Disable moving left/right
                end
                print("prone")
            elseif IsEntityPlayingAnim(ped, animationDict, animationName, 3) then
                ClearPedTasks(ped)
                EnableControlAction(0, 30, true)
                EnableControlAction(0, 31, true)
                print("standing")
            end
            Citizen.Wait(50)
        end
        Citizen.Wait(50)
    end
end)

function loadAnimDict(dict)
	while (not HasAnimDictLoaded(dict)) do
		RequestAnimDict(dict)
		Citizen.Wait(50)
	end
end
```

**Comment by MadsLeander** [comment]
Okay, so it "works" and looks ok in first person. However, once you see it from someone else's perspective then it's pretty fucked. The shooting animations are odd (pointing up into the air). However, if the animation issues are solved then there is a potential.

**Reply by blackfirefly000** [reply]
Did you try the newest version? If so, it's kinda funny that you can make it work and I can't lol. Yeah, the main issue is the animation, it's just not set up correctly to be used for shooting. I think this project is on hold if/until someone makes a proper prone shooting animation. I might look into seeing if I could get someone to make one.

**Reply by alberttheprince** [reply]
I'm discussing a potential solution with some people, but no promises on anything yet. If we can determine what the actual problem is here, might be easier to create a solution. 

This may require some digging in natives/playing around with how shooting works, I wonder if we can somehow shoot from the player (simulating shooting) while blocking the actual player ped from shooting.

**Reply by alberttheprince** [reply]
Using this native https://docs.fivem.net/natives/?_0x867654CBC7606F2C

**Reply by alberttheprince** [reply]
Idea would be: players enters a list of prone animations, we prevent the player from shooting. Get the coords of the player ped and where they're aiming, then trigger this with info from the weapon the player is holding.l when they try to shoot.

**Reply by MadsLeander** [reply]
> Did you try the newest version? If so, it's kinda funny that you can make it work and I can't lol. Yeah, the main issue is the animation, it's just not set up correctly to be used for shooting. I think this project is on hold if/until someone makes a proper prone shooting animation. I might look into seeing if I could get someone to make one.

I did use the latest code, but I had to make some changes to make it work.

**Reply by blackfirefly000** [reply]
> Idea would be: players enters a list of prone animations, we prevent the player from shooting. Get the coords of the player ped and where they're aiming, then trigger this with info from the weapon the player is holding.l when they try to shoot.

I'll take a look at some point into that

> > Did you try the newest version? If so, it's kinda funny that you can make it work and I can't lol. Yeah, the main issue is the animation, it's just not set up correctly to be used for shooting. I think this project is on hold if/until someone makes a proper prone shooting animation. I might look into seeing if I could get someone to make one.
> 
> I did use the latest code, but I had to make some changes to make it work.

Would you mind sending me your fix?

**Reply by MadsLeander** [reply]
I can't seem to find it, but for testing purposes, I removed a lot of the odd checks like this: `GetWeapontypeGroup(GetSelectedPedWeapon(ID)) == GetWeapontypeGroup(`WEAPON_SNIPER`)) and (IsPlayerFreeAiming(ped) or IsPedShooting(ped))`

**Reply by blackfirefly000** [reply]
> I can't seem to find it, but for testing purposes, I removed a lot of the odd checks like this: `GetWeapontypeGroup(GetSelectedPedWeapon(ID)) == GetWeapontypeGroup(`WEAPON_SNIPER`)) and (IsPlayerFreeAiming(ped) or IsPedShooting(ped))`

Interesting. The reason I added those is because the idea is that the player will only do the prone shooting if you were aiming or shooting a sniper, but I'm pretty sure that code isn't working correctly.

**Comment by alberttheprince** [comment]
misstrevor2ig_9b : sweep_low

Interesting animation, provides a 360 sweep aim animation.

Edit: Also, this entire dictionary basically is there for prone shooting.

misstrevor2ig_9b : reload

for example

**Reply by MadsLeander** [reply]
Isn't that from the plane mission? (Where Trevor lays on the wing of the plane and shoots all around him?) I think some of the scripts that have the prone shooting just uses a native combined with that anim set. The problem with the native is that it doesn't sync to other players... I guess "re-inventing" the same functionality would be an option.

**Comment by blackfirefly000** [comment]
Apparently, the native TaskAimGunScripted(player, `SCRIPTED_GUN_TASK_PLANE_WING`, true, true) will allow for prone shooting. Will do a re-write with RPemotes integration sometime soon(tm)

**Reply by alberttheprince** [reply]
Worth looking into, but seems to not sync?

![image](https://github.com/user-attachments/assets/2fcb5de5-a905-4eab-aa17-4c2f2d59450c)

**Reply by blackfirefly000** [reply]
I don't think this is going to be our solution. There's a [script](https://github.com/ItsAmmarB/StanceModifier) that uses it and never solved the syncing issue, and I don't think that's polished enough to push to the release version. :/

**Reply by alberttheprince** [reply]
I really think the solution is what I said earlier, in that we need to simulate shooting, with the player ped only playing an animation.

**Comment by alberttheprince** [comment]
Going to revive this thread after talking to khaosen (don't know their github) with some additional knowledge:

If we apply the  SetEnableHandcuffs it overrides shooting animations of the player ped. This allows us to put the player in an animation (lets say a prone aiming position) while holding a gun. You can then EnableControlAction shooting, which will continue to keep the ped in the animation/pose, but will allow you to fire. 

Realistically, the way you'd do "synced" shooting in this scenario is making a couple of static poses like aiming in front, aiming behind and turn the ped while in that pose, then allow shooting.

**Reply by alberttheprince** [reply]
Honestly looking at all this, I believe we should make this an entirely seperate resource that just has compatability for rpemotes. 

I think as is crouching shouldn't be part of rpemotes, but I'm going to respect the decision to keep it in since so many aplready rely on it.

**Reply by khaosen** [reply]
Hello, I have reviewed the code. These are my analysis;

As Albert has mentioned, using SetEnableHandcuffs native allow our animations to override GTA 5's native aiming animations. However, **any misusage of this will bug the character and the prop**. For example in a test script, a shield script I made, I used GetPedFlag (78) to detect if player is aiming, which apparently doesn't cover when player clicks left click without clicking right, it caused shield to go insane and rotate in extreme speeds, only to be held by the attachment to character.

Without testing, I can only speak in theory however I'd say crouch part makes it more complicated, in my test character was standing.

**Reply by khaosen** [reply]
A little note that I forgot to add, in my experiment, there was no issue with sync/fps animations.
